modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/enricher/fixed/CustomStaticTypeDeclarationEnricher.java

```diff
@@ -0,0 +1,163 @@
+import org.mule.runtime.extension.api.annotation.metadata.fixed.OutputJsonType;
+import org.mule.runtime.extension.api.annotation.metadata.fixed.OutputXmlType;
+import org.mule.runtime.extension.api.declaration.fluent.util.IdempotentDeclarationWalker;
+import org.mule.runtime.extension.api.declaration.type.annotation.CustomDefinedStaticTypeAnnotation;
+import org.mule.runtime.extension.api.loader.DeclarationEnricher;
+import org.mule.runtime.extension.api.loader.DeclarationEnricherPhase;
+import org.mule.runtime.extension.api.loader.ExtensionLoadingContext;
+import org.mule.runtime.module.extension.internal.loader.java.property.ImplementingMethodModelProperty;
+import org.mule.runtime.module.extension.internal.loader.java.property.ImplementingParameterModelProperty;
+import org.mule.runtime.module.extension.internal.loader.java.property.ImplementingTypeModelProperty;
+
+import java.io.InputStream;
+import java.lang.reflect.AnnotatedElement;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+
+/**
+ * Overrides the default output and input loaded types for the ones specified by the user using the custom static types features,
+ * such as {@link InputXmlType}, {@link OutputJsonType} and {@link InputStaticTypeResolver} for custom build types between others.
+ *
+ * @since 4.1
+ */
+public final class CustomStaticTypeDeclarationEnricher implements DeclarationEnricher {
+
+  @Override
+  public DeclarationEnricherPhase getExecutionPhase() {
+    return INITIALIZE;
+  }
+
+  @Override
+  public void enrich(ExtensionLoadingContext extensionLoadingContext) {
+    new IdempotentDeclarationWalker() {
+
+      @Override
+      protected void onOperation(WithOperationsDeclaration owner, OperationDeclaration operation) {
+        operation.getModelProperty(ImplementingMethodModelProperty.class)
+            .map(ImplementingMethodModelProperty::getMethod)
+            .ifPresent(method -> getOutputType(method).ifPresent(type -> declareCustomType(operation.getOutput(), type)));
+
+        for (ParameterDeclaration param : operation.getAllParameters()) {
+          param.getModelProperty(ImplementingParameterModelProperty.class)
+              .map(ImplementingParameterModelProperty::getParameter)
+              .ifPresent(annotated -> getInputType(annotated).ifPresent(type -> declareCustomType(param, type)));
+        }
+      }
+
+      @Override
+      protected void onSource(WithSourcesDeclaration owner, SourceDeclaration source) {
+        source.getModelProperty(ImplementingTypeModelProperty.class)
+            .map(ImplementingTypeModelProperty::getType)
+            .ifPresent(clazz -> getOutputType(clazz).ifPresent(type -> declareCustomType(source.getOutput(), type)));
+      }
+    }.walk(extensionLoadingContext.getExtensionDeclarer().getDeclaration());
+  }
+
+  private <T extends BaseDeclaration & TypedDeclaration> void declareCustomType(T declaration, MetadataType overrideType) {
```

Comment:
just to make sure, the `overrideType` is actually the type being overridden, right? is the original type that will be enriched.

Reply:
yes, Ill change the name of the param